File: algo_03_two_points/L524-m.cpp
//
// Created by LiHeng on 2020/6/3.
//

//通过删除字母匹配到字典里最长单词, 如果有多个相等长度的符合要求，返回字典序最小的那个

#include <iostream>
#include <cstring>
#include <vector>

using namespace std;
class Solution {
public:
    string findLongestWord(string s, vector<string>& d) {
        string res = "";
        for(string dd : d){
            int l1 = res.size();
            int l2 = dd.size();
            if(l1 > l2)
                continue;
            if(isSubstr(s, dd)){
                if(l1 < l2 || (l1 == l2 && res > dd))
                    res = dd;
            }
        }
        return res;
    }

    bool isSubstr(string s, string t){
        int j=0;
        for(int i=0; i<s.size(); ++i){
            if(s[i] == t[j])
                ++j;
        }
        return j==t.size();
    }
};

File: python-algorithm/algo_03_two_points/L524-m.py
#--coding:utf-8--
'''
@Time   : 2020/9/24
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''
# 和392一样

class Solution:
    def findLongestWord(self, s: str, d: List[str]) -> str:
        max_len = 0
        res = []
        for tmp_str in d:
            if self.IsSubstr(s, tmp_str):
                if len(tmp_str) > max_len:
                    res = [tmp_str]
                    max_len = len(tmp_str)
                elif len(tmp_str) == max_len:
                    res.append(tmp_str)
        if res:
            res.sort()
            return res[0]
        else:
            return ""

    def IsSubstr(self, s, t):
        s_i, t_i = 0, 0
        while s_i < len(s) and t_i < len(t):
            if s[s_i] == t[t_i]:
                s_i += 1
                t_i += 1
            else:
                s_i += 1
        return False if t_i != len(t) else True